### PR TITLE
Warn if rendering invalid element, fixes #3726

### LINKED ISSUE #3726: Warn if rendering an HTMLUnknownElement
This usually means that the user mixed up our lowercase/uppercase naming convention; see https://github.com/facebook/react/pull/3429#issuecomment-85166440.

We could add a message like

> Warning: The tag <commentBox> is unrecognized in this browser. If you meant to render a React component, start its name with an uppercase letter.

We'd check this only once per tag and only in development mode.

### PR DESCRIPTION
**Pull request by ConnorAtherton**
This PR adds a warning whenever a user attempts to render an element unrecognized by the browser. 

**TODO**
- Ensure the test suite passes (`npm test`).
- Add specific tests for `validateDOMTag` much like `validateDOMNesting`

The tests for `ReactDOMComponent` include a lot of custom web components that will now show a warning because they haven't been defined before they were used. I'm not sure how to solve this issue yet. I'm hesitant to change the existing specs and, from an initial glance, it looks like jsdom doesn't fully support the web component `customElements` property yet. @spicyj Any ideas on how to go around resolving this?

Here is a quick look at the warning:
screen shot 2016-11-23 at 08 30 05

### CODE REVIEW

File: src/renderers/dom/shared/validateDOMTag.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    if ((parentInfo && parentInfo.svgTagInScope) || instance._tag === 'svg') {
+      knownTag = !(svgTags.indexOf(instance._originalTag) === -1);
+    } else {
+      /**
+       * This will also handle custom elements instead of checking whether the tag
+       * is registered in the CustomElementRegistry.
+       *
+       * https://www.w3.org/TR/custom-elements/#custom-elements-api
+       */
+      knownTag = !(document.createElement(instance._tag) instanceof UnknownConstructor);
```
Comment (edvinerikson) [inline src/renderers/dom/shared/validateDOMTag.js]: does SSR still work with this enabled?

File: src/renderers/dom/stack/client/ReactDOMComponent.js [hunk 2]
```diff
@@ -253,6 +254,8 @@ if (__DEV__) {
     }
 
     validateDOMNesting(null, String(content), this, this._ancestorInfo);
+    validateDOMTag(this, this._ancestorInfo);
```
Comment (edvinerikson) [inline src/renderers/dom/stack/client/ReactDOMComponent.js]: I believe the changes in this file should be replicated in the Fiber component as well.

**Comment by spicyj** [comment]
It would make sense to me to not warn on any tag that has a `-` in it.

**Comment by ConnorAtherton** [comment]
Fixed in #9163

**Pull request closed without merging**